torch/__init__.py

```diff
@@ -225,6 +227,8 @@ def _load_global_deps():
 if TYPE_CHECKING:
     import torch._C as _C
 
+TensorSpec = List[Optional[str]]
```

Comment:
@jansel, @ezyang 

This is not exactly the right storage type, but I also feel torch.fx.TensorType is not exactly right either... 

I think we need something that allow user writing ('batch', Dyn, 8, Static)
'batch': dynamic, and please use 'batch' as the symbol name
Dyn: dynamic, and I don't care about the symbol name
8: static, and I know this must be 8, and compiler please error out if this is not 8
Static: static, but I don't know what's the actual value, please use the shape from sample input.

Is this over engineered?

Reply:
I'd suggest dropping support for custom names.  They seem too complicated and likely to break things down stream.  Custom names are also not composable, you might have two different codebases sharing the same name, and it is not clear if they should be implicitly merged into one or name mangled to make them different.

I'd prefer something more like debug annotations and source information, which would also be useful for shapes coming from non-inputs and a more general solution.  That seems like a separate PR though, and I wouldn't try to bundle it with this change.

Without the custom naming stuff,  I think we could just add `Static` to `torch.fx.TensorType` and get to your proposal.

---

On the UX side, there are a few options:
1) Invent some new data structure like `dynamic_args` that mirrors the args/kwargs/params.  We could pass `torch.fx.TensorType` into this data structure, or create something new like you do here.
2) Push the work of picking dynamic/static dims into a user-defined callback function, that is called once per Tensor by VariableBuilder.
3) Piggyback on Python type annotations like was done for the FX gradual shape typing work.
4) Use tensor subclassing to attach shape annotations directly to the Tensor object.

I dislike option 1 the most. I worry the data structures will be brittle (support for nesting/kwargs/params etc).  And I also want the annotations to be close to the actual data.  I think a side data structure is suboptimal UX.

2/3/4 all seem like reasonable choices to me.

In terms of easy to implement 4 might actually be easiest since you don't need to do any of the plumbing, just a small tweak to VariableBuilder.  It would also add support for non-inputs (and nested inputs) for free.

Reply:
>  Custom names are also not composable, you might have two different codebases sharing the same name, and it is not clear if they should be implicitly merged into one or name mangled to make them different.

FWIW, the behavior I'd expect is that the names are scoped only to the torch.compile expression they're associated with. So really it's just a way of conveniently specifying equality relationships between dynamic dimensions at the call site. I don't really have any bone one way or another with whether or not we should keep custom names though.

Reply:
Naming feels like a refinement / syntatic sugar once you have the core data structure designed. 

To add another option to @jansel's list above - composition feels like the best choice. Wrap a user tensor in a DynamicTensor() instead of composing, (1) to force the export system to unpack it and (2) to not deal with the annoying subclassing problems we have today for when you want to fakify a tensor subclass. 

`DynamicTensor(tensor, dynamic_dims=[], static_dims=[])` 

Also, having both static and dynamic feels like maybe too much now? Perhaps dynamic only, and assume static is fine, due to the overall specializing nature of export?

Reply:
Wrapping seems pretty similar to subclassing.  When we convert the subclass to fake tensors we can just drop/ignore the subclass and treat them like normal tensors.  The fake tensor constructor just needs to copy over the dynamic bits into the symsizes.

Reply:
The problem is now FakeTensor has to know about these, and if someone ever adds overridden methods to it, you need to do simulation or lose the expected behavior.

Reply:
Sorry, isn't DynamicTensor the subclass option 3 in https://docs.google.com/document/d/1yUik4D3LlBMGV1OIh96FEvGr6AcigrMbj8O0MXZPtdY/edit#heading=h.4oz3g6vsy0e8 which we concluded is too difficult to implement correctly (because you're obligated to propagate it through eager mode)

Reply:
> Sorry, isn't DynamicTensor the subclass option 3 in https://docs.google.com/document/d/1yUik4D3LlBMGV1OIh96FEvGr6AcigrMbj8O0MXZPtdY/edit#heading=h.4oz3g6vsy0e8 which we concluded is too difficult to implement correctly (because you're obligated to propagate it through eager mode)

The distinction is that this is (1) Export/whole_graph only (2) a meaningless notion outside the compiler stack, so no eager prop needed.

Reply:
Still need a param story tho @SherlockNoMad

Reply:
I struggle to imagine a situation when you need to mark parameter dims as dynamic.

In any case, I think it's clear we should call a meeting on this.

Reply:
@voznesenskym 

It's unclear if this is only needed for export path. In our previous discussion, we assumed that this could also useful for torch.compile() path, as having static shape would unlock some potential optimization downstream. 

Ideally, we want to have the same UX for torch.export and torch.compile, but I am fine if we end up with an export-only UX to begin with. 

I don't think we need to worry too much about dynamic parameter dims case. For export, parameters must be concrete. Otherwise, there is no way to serialize a parameter. For torch.compile, cannot think of a case where we need to change parameter shape on the fly.

Reply:
First time seeing that doc.  Based on reading it I think we should go with subclasses.   It is the only option I see that has a hope of working with graph breaks where you need to know which inputs to a graph-break-resume function are dynamic.  Dynamic shapes is *not* an export-only thing where whole graphs is an ok requirement.

Implementation wise, I don't think it will be hard.  We don't actually need to preserve the subclasses when we convert to FakeTensor and trace.  When converting DynamicTensor to FakeTensor just set the dynamic shapes on the fake tensor properly and treat it like a Tensor.

Reply:
> It's unclear if this is only needed for export path. 

No, but I think we are taking nopython=True as a firm requirement?

> Dynamic shapes is not an export-only thing where whole graphs is an ok requirement.
> It is the only option I see that has a hope of working with graph breaks

I thought we had decided firmly against propagation across eager? cc: @Chillee

Reply:
We'll talk about this tomorrow.